### ISSUE TITLE
renderDot does not trigger callback

### ISSUE THREAD

**Issue by GrahamHannington** [magjac/d3-graphviz#14]
This works (renders the graph in `data`):

    d3.select("#graph").graphviz().renderDot(data);

So does this (as above, with an anonymous callback function), except that I don't get a console message; renderDot does not trigger the callback:

    d3.select("#graph").graphviz().renderDot(data, function(){console.log("Ended")});

I'm interested in using a callback function because I want to zoom the graph to fit the viewport. That's what I want to do in the callback (although, I'm still working out the "zoom to fit" function).

I'm using latest Google Chrome on Windows 10.

In case it's related, the Chrome console shows the following error (which doesn't seem to affect the rendering):

Invalid asm.js: Function definition doesn't match use - viz.js:37

Software versions:

    <script src="//d3js.org/d3.v4.min.js"></script>
    <script src="https://unpkg.com/viz.js@1.8.0/viz.js" type="javascript/worker"></script>
    <script src="https://unpkg.com/d3-graphviz@1.0.1/build/d3-graphviz.min.js"></script>

**Comment by magjac** [comment]
I'll look into this issue in the nearest couple of days. I the meantime, you can use
`d3.select("#graph").graphviz().renderDot(data).on("end", function() {console.log("Ended")});`

See this [example](https://github.com/magjac/d3-graphviz/blob/master/examples/demo.html#L26).

Regarding the "zoom to fit" function, you might want to use the `.attributer()` method instead of a callback. See https://github.com/magjac/d3-graphviz/issues/7

Thanks for submitting this issue. If you are using this library, don't forget to "Watch" it to stay tuned for updates or even "Star" it if you like it.

You can also use [Stack Overflow tags d3.js & graphviz](https://stackoverflow.com/tags/d3.js+graphviz) to ask for help. Although I make an effort to assist everyone that asks, I am not always available to provide help promptly or directly.

**Comment by GrahamHannington** [comment]
Thanks for the quick reply. I should have mentioned this in my initial report: `.on("end")...` doesn't work for me, either. Makes me think I'm doing something wrong, but the DOT renders just fine.

**Comment by GrahamHannington** [comment]
Re: "zoom to fit"

I can see that the D3 V4 dynamic (mouse-driven) panning and zooming dynamically updates the `transform` attribute of the SVG `g` element that was generated by d3-graphviz:

    <div id="graph">
      <svg width="2445pt" height="1538pt" viewBox="0.00 0.00 2445.00 1538.36" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
        <g id="graph0" class="graph" transform="translate(43.842239806463624,1357.7823048955277) scale(0.8705505632961241)">

This makes me think that any "zoom to fit" function should update that `transform` attribute of the `g` element, too, to "play nice" with the dynamic D3 panning and zooming. (Although I'm scratching my head over how to calculate an appropriate value to fit the `g` to the viewport. I'll be reading some SVG documentation tonight.)

However, the `attributer` example you pointed me to (thanks!) targets the `svg` element (parent of that `g` element).

Your thoughts?

**Comment by magjac** [comment]
Maybe I misunderstood. I meant an initial zoom to fit the generated graph, from which the user can zoom in or out, but now  I realize that you probably meant a user-initiated function that fits the graph regardless of its original size and current zoom level.

I think however, that they can be combined. I'll look into it late tonight CET (I'm coding for money during the day).

**Comment by magjac** [comment]
On my phone so I can’t test it, but the callback/ event seems to fire only when a transition is used. As a workaround, try using a dummy transition with a duration of 0.

And yes, the invalid asm.js is unrelated. It’s a known viz.js issue.

**Comment by magjac** [comment]
Tested this workaround ok:
```
d3.select("#graph").graphviz()
    .transition("dummy")
    .renderDot('digraph  {a -> b}', function () {
        console.log('Callback');
    })
    .on("end", function () {
        console.log('"end" event');
    });
```

**Comment by magjac** [comment]
I've pushed a fix for this bug in https://github.com/magjac/d3-graphviz/commit/1277bd2ba53c1a4736fd1e980368a09553ccffce. Do you need a new release or do you install locally?

Also, can I ask you to post a question on [Stack Overflow using the tags d3.js and graphviz](https://stackoverflow.com/tags/d3.js+graphviz) if you still need help with the  "zoom to fit" function? It's a better place to answer questions. I've started to work on it, but now I must call it a day.

**Comment by GrahamHannington** [comment]
I'm sincerely grateful for your help.

A new release would be great. Right now, I'm being "lazy", and just doing this:

    <script src="https://unpkg.com/d3-graphviz@1.0.1/build/d3-graphviz.min.js"></script>

I'm concentrating on tweaking the program that generates the DOT file.

Yes, if I can't work it out myself in the next day or so, I'll post a question on Stack Overflow about the "zoom to fit" function.

> I meant an initial zoom to fit the generated graph ... I think however, that they can be combined.

Yes, I had in mind a single function that could be called at initial "render time", and then called later whenever the user clicks a "Fit" button.

**Comment by magjac** [comment]
Released in [v1.1.0](https://github.com/magjac/d3-graphviz/releases/tag/v1.1.0)

**Comment by GrahamHannington** [comment]
Thank you! I'm now using a `.renderDot` callback successfully in v1.1.0
